interceptors/logging/logging.go

```diff
@@ -161,30 +164,42 @@ NextAddField:
 // If there are no fields in the context, it returns an empty Fields value.
 // Extracted fields are useful to construct your own logger that has fields from gRPC interceptors.
 func ExtractFields(ctx context.Context) Fields {
-	t, ok := ctx.Value(fieldsCtxMarkerKey).(Fields)
+	t, ok := ctx.Value(fieldsCtxMarkerKey).(*fieldsCtxValue)
 	if !ok {
 		return nil
 	}
-	n := make(Fields, len(t))
-	copy(n, t)
+	n := make(Fields, len(t.fields))
+	copy(n, t.fields)
 	return n
 }
 
-// InjectFields allows adding fields to any existing Fields that will be used by the logging interceptor or can be
+// InjectFields returns a new context with merged fields that will be used by the logging interceptor or can be
 // extracted further in ExtractFields.
 // For explicitness, in case of duplicates, the newest field occurrence wins. This allows nested components to update
 // popular fields like grpc.component (e.g. server invoking gRPC client).
 //
 // Don't overuse mutation of fields to avoid surprises.
 func InjectFields(ctx context.Context, f Fields) context.Context {
-	return context.WithValue(ctx, fieldsCtxMarkerKey, f.WithUnique(ExtractFields(ctx)))
+	return context.WithValue(ctx, fieldsCtxMarkerKey, &fieldsCtxValue{fields: f.WithUnique(ExtractFields(ctx))})
 }
 
 // InjectLogField is like InjectFields, just for one field.
 func InjectLogField(ctx context.Context, key string, val any) context.Context {
 	return InjectFields(ctx, Fields{key, val})
 }
 
+// AddFields updates the fields already in the context that will be used by the logging interceptor or can be
+// extracted further in ExtractFields. For explicitness, in case of duplicates, the newest field occurrence wins.
+//
+// This function is not safe to call concurrently.
```

Comment:
I'm sympathetic to the use case, but I do kinda hate the way we have to implement it. I don't feel strongly enough to refuse this functionality, but maybe we could add a note to the comment to say that in general, `InjectFields` should be preferred over `AddFields`?
```suggestion
// AddFields updates the fields already in the context that will be used by the logging interceptor or can be
// extracted further in ExtractFields. For explicitness, in case of duplicates, the newest field occurrence wins.
//
// InjectFields should be used instead of AddFields where possible, as it does not require mutating the values
// already in the context.
//
// This function is not safe to call concurrently.
```